codegen/src/api/metadata.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is part of subxt.
+//
+// subxt is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// subxt is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with subxt.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Wrapper to obtain unique deterministic hashed identifiers from portable type ids.
+//!
+//! # Note
+//!
+//! Used to determine API compatibility between generated interface and dynamic metadata.
+
+use codec::Encode;
+use frame_metadata::{
+    RuntimeMetadataLastVersion,
+    StorageEntryType,
+};
+use scale_info::{
+    form::PortableForm,
+    Field,
+    TypeDef,
+    Variant,
+};
+use std::{
+    collections::{
+        HashMap,
+        HashSet,
+    },
+    sync::Mutex,
+};
+
+use lazy_static::lazy_static;
+
+#[derive(Clone, Debug)]
+pub struct MetadataHashable {
+    metadata: RuntimeMetadataLastVersion,
+}
+
+#[repr(u8)]
+enum MetadataHashableIDs {
+    Field,
+    Variant,
+    TypeDef,
+    Type,
+    Pallet,
+}
+
+impl MetadataHashable {
+    pub fn new(metadata: RuntimeMetadataLastVersion) -> Self {
+        Self { metadata }
+    }
+
+    pub fn metadata(&self) -> &RuntimeMetadataLastVersion {
+        &self.metadata
+    }
+
+    fn hash(bytes: &[u8]) -> [u8; 32] {
+        sp_core::hashing::sha2_256(bytes)
+    }
+
+    fn get_field_uid(
+        &self,
+        field: &Field<PortableForm>,
+        set: &mut HashSet<u32>,
+    ) -> [u8; 32] {
+        let mut bytes = vec![MetadataHashableIDs::Field as u8];
+
+        if let Some(name) = field.name() {
+            bytes.extend(name.as_bytes());
+        }
+        if let Some(ty_name) = field.type_name() {
+            bytes.extend(ty_name.as_bytes());
+        }
+        bytes.extend(self.get_type_uid_internal(field.ty().id(), set));
+
+        MetadataHashable::hash(&bytes)
+    }
+
+    fn get_variant_uid(
+        &self,
+        var: &Variant<PortableForm>,
+        set: &mut HashSet<u32>,
+    ) -> [u8; 32] {
+        let mut bytes = vec![MetadataHashableIDs::Variant as u8];
+
+        bytes.extend(var.name().as_bytes());
+        for field in var.fields() {
+            bytes.extend(self.get_field_uid(field, set));
+        }
+
+        MetadataHashable::hash(&bytes)
+    }
+
+    fn get_type_def_uid(
```

Comment:
Nice! This approach makes sense to me :)

Reply:
Thanks :D